File: back-end/src/main/java/ccu/four/backend/models/note/TeamNote.java
package ccu.four.backend.models.note;

import ccu.four.backend.enums.Specialty;
import ccu.four.backend.models.Team;
import ccu.four.backend.requestBodies.TeamNoteRequestBody;
import com.fasterxml.jackson.annotation.JsonBackReference;

import javax.persistence.*;
import java.util.ArrayList;
import java.util.List;

@Entity
public class TeamNote extends Note {

    // TODO: Make nullable false, it's true for testing purposes.
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "teamId", nullable = true)
    @JsonBackReference
    private Team team;

    @ElementCollection
    private List<Specialty> specialties;

    private boolean isPublic;

    public TeamNote() {
        super();
        this.specialties = new ArrayList<>();
    }

    public TeamNote(Team team) {
        super();
        this.specialties = new ArrayList<>();
        this.team = team;
    }

    public TeamNote(Team team, String header, String content) {
        super(header, content);
        this.team = team;
        this.specialties = new ArrayList<>();
    }

    public TeamNote(TeamNoteRequestBody teamNoteRequestBody) {
        super();
        super.setHeader(teamNoteRequestBody.getHeader());
        super.setContent(teamNoteRequestBody.getContent());
        super.setDate(teamNoteRequestBody.getDate());
        super.setId(teamNoteRequestBody.getId());
        super.setUser(teamNoteRequestBody.getUser());
        this.setPublic(teamNoteRequestBody.getPublic());
        this.specialties = new ArrayList<>(teamNoteRequestBody.getSpecialties());
    }

    public void addSpecialty(Specialty specialty) {
        specialties.add(specialty);
    }

    public void addSpecialties(Specialty... args) {
        specialties.addAll(List.of(args));
    }

    public void removeSpecialty(Specialty specialty) {
        specialties.remove(specialty);
    }

    public Team getTeam() {
        return team;
    }

    public List<Specialty> getSpecialties() {
        return specialties;
    }

    public void setPublic(boolean aPublic) {
        isPublic = aPublic;
    }

    public boolean isPublic() {
        return isPublic;
    }

    public void setTeam(Team team) {
        this.team = team;
    }
}

File: front-end/src/app/models/notes/teamNote.ts
import {Team} from '../team';
import {Specialty} from "../user";
import {Note} from "./note";

export class TeamNote extends Note {
  public get isPublic(): boolean {
    return this.public;
  }

  public set isPublic(aPublic: boolean) {
    this.public = aPublic;
  }

  public public: boolean;
  public specialties: Specialty[] = [];
  public team: Team;

  public static readonly path = "team";

  public getPath(): string {
    return TeamNote.path;
  }

  public getParentID(): string {
    return this.team.id.toString();
  }

  public getIdentifier(): string | number {
    return this.team.id;
  }

  public getListOfSpecialties(): Specialty[] {
    return [...this.specialties];
  }

  public static getEmpty(): TeamNote {
    return new TeamNote()
  }

  public getEmpty(): TeamNote {
    return TeamNote.getEmpty();
  }
}
